src/weekly/mod.rs

```diff
@@ -0,0 +1,337 @@
+use std::time::Duration;
+
+/// Whole week duration, in minutes.
+pub(crate) const MAX_WEEKLY_MINS: u32 = 7 * 24 * 60;
+
+/// A weekly point in time, as minutes since beginning of week (Monday 00:00).
+pub(crate) type MinuteInWeek = u32;
+
+/// Calendar for periodic time-windows, recurring on weekly basis.
+#[derive(Clone, Debug)]
+pub struct WeeklyCalendar {
+    /// An immutable set of (possibly overlapping) daily intervals.
+    windows: IntervalTree<MinuteInWeek, DailyWindow>,
+}
+
+impl WeeklyCalendar {
+    /// Create a weekly calendar from a vector of daily windows.
+    pub fn new(input: Vec<DailyWindow>) -> Self {
+        let intervals = input
+            .into_iter()
+            .map(|win| Element::from((win.range_weekly_minutes(), win)));
+
+        Self {
+            windows: IntervalTree::from_iter(intervals),
+        }
+    }
+
+    /// Return whether datetime is contained in this weekly calendar.
+    pub fn contains_datetime(&self, datetime: &DateTime<Utc>) -> bool {
+        let timepoint = utils::datetime_as_weekly_minute(datetime);
+        self.windows.query_point(timepoint).count() > 0
+    }
+
+    /// Return total length of all daily windows in the calendar.
+    ///
+    /// In case of overlapping windows, total length can be larger than the
+    /// actual amount of weekly minutes in the calendar.
+    pub fn total_length_minutes(&self) -> u64 {
+        self.windows.iter().fold(0u64, |len, win| {
+            len.saturating_add(win.value.length_minutes().into())
+        })
+    }
+
+    /// Return all daily windows (if any) which contain a given datetime.
+    #[cfg(test)]
+    pub fn containing_windows(&self, datetime: &DateTime<Utc>) -> Vec<&DailyWindow> {
+        let timepoint = utils::datetime_as_weekly_minute(datetime);
+        self.windows
+            .query_point(timepoint)
+            .map(|elem| &elem.value)
+            .collect()
+    }
+
+    // TODO(lucab): compute and expose coalesced length.
+}
+
+impl Serialize for WeeklyCalendar {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        use serde::ser::SerializeSeq;
+
+        let len = self.windows.iter().count();
+        let mut seq = serializer.serialize_seq(Some(len))?;
+        for interval in self.windows.iter() {
+            seq.serialize_element(&interval.value)?;
+        }
+        seq.end()
+    }
+}
+
+impl Default for WeeklyCalendar {
+    fn default() -> Self {
+        Self::new(vec![])
+    }
+}
+
+/// Timespan with a fixed duration, recurring on weekly basis.
+#[derive(Clone, Debug, Eq, Serialize)]
+pub struct DailyWindow {
+    weekday: chrono::Weekday,
+    start_hour: u8,
+    start_minute: u8,
+    length: Duration,
+}
+
+impl DailyWindow {
+    /// Parse a timespan into daily windows.
+    pub fn parse_timespan(
+        start_day: chrono::Weekday,
+        start_hour: u8,
+        start_minute: u8,
+        length: Duration,
+    ) -> Fallible<Vec<Self>> {
```

Comment:
That was in my initial implementation, but it was a bit too naïf thus failing on corner-cases.

In particular, a weekly calendar is a [finite field of order 7](https://en.wikipedia.org/wiki/Finite_field). Perhaps not very intuitive, but all the logic here has to obey [modular arithmetic](https://en.wikipedia.org/wiki/Modular_arithmetic) in that field. (same goes for hours and minutes too, really)

In other words, it has to account for the fact that a timespan can start on Sunday night and end on Monday morning, without jumping back in time :)

Splitting into multiple daily windows here makes the rest of the intervals logic easier to deal with, as it can simply ignore all modular details and just use integer arithmetic when comparing `MinuteInWeek`.

Reply:
Right, that makes sense. I guess we could split just the intervals which cross week boundaries into two. But meh... I see the appeal in being consistent.

Reply:
Uh, good point, I somehow assumed it was easier to handle interval windows on daily boundaries but that's not a given. I'll try to quickly sketch your alternative next and check if it is easier to follow.

Reply:
I reworked the logic here to use week boundaries, and the final result is indeed easier to read. I've stick to that and pushed this new version, thanks for the suggestion!